kirby/supervisor/__init__.py

```diff
@@ -1,46 +1,107 @@
-from kirby.supervisor.scheduler import Scheduler
+from kirby.supervisor.scheduler import Scheduler, MessageType
 from kirby.supervisor.executor.runner import Runner
 from kirby.supervisor.executor.arbiter import Arbiter
 
 logger = logging.getLogger(__name__)
 
+USE_TLS = getenv("KAFKA_USE_TLS", type=bool, default=False)
+JOB_OFFERS_TOPIC_NAME = getenv(
+    "KIRBY_TOPIC_JOB_OFFERS", type=str, default=".kirby.job-offers"
+)
+RUNNING_DEAMONS_TOPIC_NAME = getenv(
+    "KIRBY_TOPIC_JOB_OFFERS", type=str, default=".kirby.job-offers"
+)
 
-def run_supervisor(name, window, wakeup, nb_runner, nb_arbiter):
+
+def send_arbiters_job_if_needed(
+    queue_list_arbiters, arbiters, last_update_message
+):
+    if last_update_message:
+        start = last_update_message.timestamp
+    else:
+        start = earliest_kafka_date()
+
+    end = datetime.datetime.utcnow()
+
+    # Search for new_update_message
+    new_update_message = None
+    for message in queue_list_arbiters.rewind(earlier=start, latest=end):
+        if message.headers["type"] == MessageType.UPDATE.value:
+            new_update_message = message
+            break
+
+    if new_update_message:
+        last_update_message = new_update_message
+        for arbiter in arbiters:
+            queue_list_arbiters.append(
+                arbiter.job.json_repr(),
+                headers={"type": MessageType.JOB.value},
+            )
+
+    return last_update_message
+
+
+def run_supervisor(name, window, wakeup, nb_runner):
     server = Redis()
-    queue = Queue(
-        name=getenv(
-            "KIRBY_TOPIC_JOB_OFFERS", type=str, default=".kirby.job-offers"
-        ),
-        use_tls=getenv("KAFKA_USE_SSL", type=bool, default=False),
+
+    queue_job_offers_arbiters = Queue(
+        name=JOB_OFFERS_TOPIC_NAME, use_tls=USE_TLS, group_id=".kirby.arbiters"
+    )
+    queue_job_offers_runners = Queue(
+        name=JOB_OFFERS_TOPIC_NAME, use_tls=USE_TLS, group_id=".kirby.runners"
+    )
+    queue_running_deamons = Queue(
+        name=RUNNING_DEAMONS_TOPIC_NAME, use_tls=USE_TLS
+    )
+
+    scheduler = Scheduler(
+        queue_job_offers=queue_job_offers_arbiters,
+        queue_running_deamons=queue_running_deamons,
+        wakeup=wakeup,
     )
-    scheduler = Scheduler(queue=queue, wakeup=wakeup)
 
     for i in range(nb_runner):
-        Runner(queue=scheduler.queue)
-    for i in range(nb_arbiter):
-        Arbiter(queue=scheduler.queue)
+        Runner(queue_job_offers_runners)
 
+    arbiters = []
+    last_update_message = None
     with Election(identity=name, server=server, check_ttl=window) as me:
         while True:
             checkpoint = perf_counter()
             if me.is_leader():
                 content = scheduler.fetch_jobs()
                 if content is not None:
                     jobs = scheduler.parse_jobs(content)
+                    running_arbiters_job = scheduler.get_running_arbiters_job(
+                        nb_supervisors
```

Comment:
We need to define this variable first?

Reply:
Yes, the commit I've made was for @nicksjl-adimian to see the advancement
The behavior will soon change completely, motivated by our discussion